migrations/admin_events.go

```diff
@@ -0,0 +1,151 @@
+package migrations
+
+import (
+	"fmt"
+
+	"github.com/mongodb/anser"
+	"github.com/mongodb/anser/db"
+	"github.com/mongodb/anser/model"
+	"github.com/pkg/errors"
+	"gopkg.in/mgo.v2/bson"
+)
+
+const (
+	eventCollection       = "event_log"
+	adminDataType         = "ADMIN"
+	eventTypeKey          = "e_type"
+	eventTypeBanner       = "BANNER_CHANGED"
+	eventTypeTheme        = "THEME_CHANGED"
+	eventTypeServiceFlags = "SERVICE_FLAGS_CHANGED"
+	eventTypeValueChanged = "CONFIG_VALUE_CHANGED"
+)
+
+type eventDataOld struct {
+	ResourceType string            `bson:"r_type"`
+	User         string            `bson:"user"`
+	OldVal       string            `bson:"old_val"`
+	NewVal       string            `bson:"new_val"`
+	OldFlags     eventServiceFlags `bson:"old_flags"`
+	NewFlags     eventServiceFlags `bson:"new_flags"`
+}
+
+type eventServiceFlags struct {
+	TaskDispatchDisabled         bool `bson:"task_dispatch_disabled"`
+	HostinitDisabled             bool `bson:"hostinit_disabled"`
+	MonitorDisabled              bool `bson:"monitor_disabled"`
+	NotificationsDisabled        bool `bson:"notifications_disabled"`
+	AlertsDisabled               bool `bson:"alerts_disabled"`
+	TaskrunnerDisabled           bool `bson:"taskrunner_disabled"`
+	RepotrackerDisabled          bool `bson:"repotracker_disabled"`
+	SchedulerDisabled            bool `bson:"scheduler_disabled"`
+	GithubPRTestingDisabled      bool `bson:"github_pr_testing_disabled"`
+	RepotrackerPushEventDisabled bool `bson:"repotracker_push_event_disabled"`
+	CLIUpdatesDisabled           bool `bson:"cli_updates_disabled"`
+	GithubStatusAPIDisabled      bool `bson:"github_status_api_disabled"`
+}
+
+type eventDataNew struct {
+	ResourceType string           `bson:"r_type"`
+	User         string           `bson:"user"`
+	Section      string           `bson:"section"`
+	Changes      configDataChange `bson:"changes"`
+}
+
+type configDataChange struct {
+	Before interface{} `bson:"before"`
+	After  interface{} `bson:"after"`
+}
+
+type dbSettings struct {
+	Banner      string `bson:"banner"`
+	BannerTheme string `bson:"banner_theme"`
+}
+
+func adminEventRestructureGenerator(env anser.Environment, dbName string, limit int) (anser.Generator, error) {
+	const migrationName = "admin_event_restructure"
+
+	if err := env.RegisterManualMigrationOperation(migrationName, makeAdminEventMigration(dbName)); err != nil {
+		return nil, err
+	}
+
+	opts := model.GeneratorOptions{
+		NS: model.Namespace{
+			DB:         dbName,
+			Collection: eventCollection,
+		},
+		Limit: limit,
+		Query: db.Document{
+			"r_id":        "",
```

Comment:
As discussed, this will run every time. You can query on the presence of "old_vals" if you want. Another possibility that keeps this indexed would be is set r_id.

Reply:
done